File: flink-sample-java/src/main/java/KeyedTransformations.java
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import util.SensorReading;
import util.SensorSource;
import util.SensorTimeAssigner;

public class KeyedTransformations {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
        env.getConfig().setAutoWatermarkInterval(1000L);

        DataStream<SensorReading> readings = env
                .addSource(new SensorSource())
                .assignTimestampsAndWatermarks(new SensorTimeAssigner());
        KeyedStream<SensorReading, String> keyed = readings.keyBy(r -> r.id);
        DataStream<SensorReading> maxTempPerSensor = keyed
                .reduce((r1, r2) -> {
                    if (r1.temperature > r2.temperature) {
                        return r1;
                    } else {
                        return r2;
                    }
                });
        maxTempPerSensor.print();
        env.execute();
    }
}

File: flink-datastream-1.7.1/src/main/scala/scalalang/KeyedTransformations.scala
package scalalang

import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.api.scala._
import scalalang.source.SensorSource
import scalalang.timestamp.SensorTimeAssigner

object KeyedTransformations {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    env.getConfig.setAutoWatermarkInterval(1000L)

    val readings = env
      .addSource(new SensorSource)
      .assignTimestampsAndWatermarks(new SensorTimeAssigner)

    val keyed = readings.keyBy(_.id)
    val maxTempPerSensor = keyed.reduce((r1, r2) => {
      if (r1.temperature > r2.temperature) r1 else r2
    })

    maxTempPerSensor.print()
    env.execute()
  }
}
